chore(flags): Add new debugging property `$feature_flag_bootstrapped_response` and `$feature_flag_bootstrapped_payload` to `$feature_flag_called` event

havenbarnes opened the pull request:
## Problem

When SDK users make use of feature flag bootstrapping, the response shown in the $feature_flag_called event can show different values than what would have been fetched remotely. This can lead to confusion when reviewing flag value breakdown results, for example making it seem as if a flag that is turned completely to `true` was evaluated to `false`. These two new event properties are mostly useful for debugging customer issues, but could also be used in product to show users that the response shown came from bootstrapped values, and not /decide

## Changes

<!-- What is changed and what information would be useful to a reviewer? -->

Nothing functional, just additional event properties

## Release info Sub-libraries affected

### Bump level

<!-- Please mark what level of change this is. -->

- [ ] Major
- [x] Minor
- [ ] Patch

### Libraries affected

<!-- Please mark which libraries will require a version bump. -->

- [ ] All of them
- [x] posthog-web
- [ ] posthog-node (no server-side bootstrapping)
- [x] posthog-react-native

### Changelog notes

<!-- Add notes here that should be added to the changelogs. -->

- Add new debugging property `$feature_flag_bootstrapped_response` and `$feature_flag_bootstrapped_payload` to `$feature_flag_called` event

havenbarnes commented:
@dmarticus I'm thinking it may be smart to add an additional property here as well, indicated whether flags have been fetched yet or not at the time of the flag check. Something like `$feature_flags_fetched: boolean`. WDYT?

dmarticus commented:
> @dmarticus I'm thinking it may be smart to add an additional property here as well, indicated whether flags have been fetched yet or not at the time of the flag check. Something like `$feature_flags_fetched: boolean`. Otherwise, we don't have a clear way to know if `$feature_flag_response` matches `$feature_flag_bootstrapped_response` because bootstrapping was done correctly, or if the flags have just been fetched already. WDYT?

I love this idea; seems like it couldn't hurt to add this type of metadata to our flags requests – maybe something like `used_boostrap_value` is more clear than `feature_flags_fetched`, if the goal is to track that bootstrapping was done correctly.  And I think that could be done in a subsequent PR.  But I like this!

dmarticus reviewed (approved):
stamped with a comment!

marandaneto commented on posthog-react-native/CHANGELOG.md:
```diff
@@ -1,5 +1,9 @@
 # Next
 
+# 3.4.1 - 2024-11-27
```
Should this be a minor since it adds a new feature?

havenbarnes replied:
@marandaneto  It's technically just for debugging ability, no plans to use in product right now. You know better here, so let me know what you think!

dmarticus replied:
eh I think @marandaneto has a good point since it's kinda going to be used in our product (once those events get propagated to the payload, all users will have access to it).  Useful for debugging on our end, sure, but could also be handy for customers I imagine.

marandaneto replied:
your call here, just food for though, the patch is just for bug fixes usually (semver).

havenbarnes replied:
Makes sense for traditional semver, updated to minor bump!

marandaneto commented:
@havenbarnes would be cool to add a test case for bootstrap flags and payloads, check the file `posthog.featureflags.spec.ts`

havenbarnes commented:
Thanks @marandaneto , added some tests. Also added the `$used_bootstrap_value` property discussed above. I'm imitating same behavior as `this.instance.decideEndpointWasHit` from `posthog-js` (https://github.com/PostHog/posthog-js/blob/bootstrapped-flag-evt-property/src/posthog-featureflags.ts#L295). I'll wait for your approval!

marandaneto commented on posthog-react-native/CHANGELOG.md:
```diff
@@ -1,13 +1,17 @@
 
 1. feat: automatically mask out user photos and sandboxed views like photo picker (iOS Only)
-  1. To disable masking set `maskAllSandboxedViews` and `maskPhotoLibraryImages` to false
+1. To disable masking set `maskAllSandboxedViews` and `maskPhotoLibraryImages` to false
```
I believe those changes (the rest of the file) are unintentional since its a nested bullet list, please revert

havenbarnes replied:
Weird my VS code was doing this, reverted

marandaneto commented on posthog-web/CHANGELOG.md:
```diff
@@ -1,5 +1,11 @@
+
+## Changed
+
+1. Add new debugging property `$feature_flag_bootstrapped_response` and `$feature_flag_bootstrapped_payload` to `$feature_flag_called` event
```
we can also mention `$used_bootstrap_value`

marandaneto commented on posthog-web/CHANGELOG.md:
```diff
@@ -1,5 +1,11 @@
 # Next
 
+# 3.1.1 - 2024-11-27
```
this should be a minor as well similarly to RN

marandaneto commented on posthog-core/src/index.ts:
```diff
@@ -1252,6 +1258,14 @@ export abstract class PostHogCore extends PostHogCoreStateless {
+        $feature_flag_bootstrapped_payload: this.getPersistedProperty<PostHogDecideResponse['featureFlagPayloads']>(
+          PostHogPersistedProperty.BootstrapFeatureFlagPayloads
+        )?.[key],
+        // If we haven't yet received a response from the /decide endpoint, we must have used the bootstrapped value
+        $used_bootstrap_value: !this.getPersistedProperty(PostHogPersistedProperty.DecideEndpointWasHit),
```
I think it's ok having this but you could also just compare the flag (the cached flag - likely not there if decide isn't called, and the value from the bootstrap flags), so you know if it's being used from the bootstrap or not, right?
Since using storage is not super fast, the less we need the better, your call here, but if possible I'd evaluate this at runtime.

havenbarnes replied:
Are you referring to `PostHogPersistedProperty.FeatureFlags` or a platform specific cache?

I was hoping we could avoid this as well, but we store bootstrapped flags in `PostHogPersistedProperty.FeatureFlags` upon initialization, so it would already be there

havenbarnes commented:
> @havenbarnes I think you can merge/release this, that would help debugging #314 o/

Thanks @marandaneto , i was hoping to wait for to get unblocked https://github.com/PostHog/posthog-js/pull/1567 but obviously this will be helpful regardless. If I can't get it unblocked today I'll just merge this
